File: code_snippets/ts/src/operate/invocations/ingress.ts
import {greetCounterObject, greeterService} from "./utils";
import {SendOpts} from "@restatedev/restate-sdk-clients";
import {myWorkflow} from "../../concepts/invocations/utils";

// <start_clients_import>
import * as clients from "@restatedev/restate-sdk-clients";
// <end_clients_import>

const myPlainTSFunction = async () => {
    // <start_rpc_call_node>
        const rs = clients.connect({url: "http://localhost:8080"});
        const greet = await rs.serviceClient(greeterService)
            .greet({greeting: "Hi"});

        const count = await rs.objectClient(greetCounterObject, "Mary")
            .greet({greeting: "Hi"});

        // You cannot **submit** workflows via request-response with the clients.
        // You can call the other handlers of workflows via request-response.
        const response = await rs.workflowClient(myWorkflow, "wf-id-1")
            .myOtherHandler({input: "Hi"});
    // <end_rpc_call_node>
}

const myPlainTSFunction2 = async () => {
    // <start_one_way_call_node>
    const rs = clients.connect({url: "http://localhost:8080"});
    await rs.serviceSendClient(greeterService)
        .greet({greeting: "Hi"});

    await rs.objectSendClient(greetCounterObject, "Mary")
        .greet({greeting: "Hi"});

    await rs.workflowClient(myWorkflow, "wf-id-1")
        .workflowSubmit({input: "Hi"});
    // You cannot yet do one-way calls to other handlers of the workflows
    // <end_one_way_call_node>
}

const myPlainTSFunction3 = async () => {
    // <start_delayed_call_node>
    const rs = clients.connect({url: "http://localhost:8080"});
    await rs.serviceSendClient(greeterService)
        .greet({greeting: "Hi"}, SendOpts.from({ delay: 1000 }));

    await rs.objectSendClient(greetCounterObject, "Mary")
        .greet({greeting: "Hi"}, SendOpts.from({ delay: 1000 }));

    // You cannot call any workflow handlers via delayed calls with the clients.
    // <end_delayed_call_node>
}

const servicesIdempotent = async () => {
    const request = {greeting: "Hi"}
    // <start_service_idempotent>
    const rs = clients.connect({url: "http://localhost:8080"});
    await rs.serviceSendClient(greeterService)
        // withClass highlight-line
        .greet(request, SendOpts.from({ idempotencyKey: "abcde" }));
    // <end_service_idempotent>
}

const servicesAttach = async () => {
    const request = {greeting: "Hi"}
    // <start_service_attach>
    const rs = clients.connect({url: "http://localhost:8080"});
    // Send a message
    const handle = await rs.serviceSendClient(greeterService)
        .greet(request, SendOpts.from({ idempotencyKey: "abcde" }));

    // ... do something else ...

    // Attach later to retrieve the result
    // withClass highlight-line
    const response = await rs.result(handle);
    // <end_service_attach>
}

const workflowAttach = async () => {
    // <start_workflow_attach>
    const rs = clients.connect({url: "http://localhost:8080"});

    // Submit the workflow
    const handle = await rs.workflowClient(myWorkflow, "wf-id-1")
        .workflowSubmit({input: "Hi"});

    // ... do something else ...

    // Attach by using the handle:
    // withClass highlight-line
    const result = await rs.result(handle);

    // Or, attach by using the workflow ID (from another service):
    // withClass highlight-line
    const result2 = await rs.workflowClient(myWorkflow, "wf-id-1").workflowAttach();
    // <end_workflow_attach>
}

const workflowPeek = async () => {
    // <start_workflow_peek>
    const rs = clients.connect({url: "http://localhost:8080"});

    // Peek at the output by using the workflow ID (from another service):
    // withClass highlight-line
    const output = await rs.workflowClient(myWorkflow, "wf-id-1").workflowOutput();
    // If the workflow is ready, do something with the result
    if (output.ready) {
        console.log(output.result);
    }
    // <end_workflow_peek>
}

File: code_snippets/java/src/main/java/operate/invocations/Ingress.java
package operate.invocations;

import dev.restate.sdk.Context;
import dev.restate.sdk.JsonSerdes;
import dev.restate.sdk.client.Client;
import dev.restate.sdk.client.SendResponse;
import dev.restate.sdk.common.Output;
import develop.MyWorkflowClient;

import java.time.Duration;

import static dev.restate.sdk.client.CallRequestOptions.DEFAULT;

public class Ingress {

    public void myJavaHandler(Context ctx) {
        // <start_rpc_java>
        Client rs = Client.connect("http://localhost:8080");
        String greet = GreeterServiceClient.fromClient(rs)
            .greet("Hi");

        int count = GreetCounterObjectClient.fromClient(rs, "Mary")
            .greet("Hi");
        // <end_rpc_java>
    }

    public void myOneWayCallHandler(Context ctx) {

        // <start_one_way_call_java>
        Client rs = Client.connect("http://localhost:8080");
        GreeterServiceClient.fromClient(rs)
                .send()
                .greet("Hi");

        GreetCounterObjectClient.fromClient(rs, "Mary")
                .send()
                .greet("Hi");

        MyWorkflowClient.fromClient(rs, "wf-id-1")
                .submit("input");
        // <end_one_way_call_java>
    }

    public void myDelayedOneWayCallHandler(Context ctx) {
        // <start_delayed_call_java>
        Client rs = Client.connect("http://localhost:8080");
        GreeterServiceClient.fromClient(rs)
            .send(Duration.ofMillis(1000))
            .greet("Hi");

        GreetCounterObjectClient.fromClient(rs, "Mary")
            .send(Duration.ofMillis(1000))
            .greet("Hi");
        // <end_delayed_call_java>

    }

    public void idempotentInvoke(){
        // <start_service_idempotent>
        Client rs = Client.connect("http://localhost:8080");
        GreetCounterObjectClient.fromClient(rs, "Mary")
            .send()
            // withClass highlight-line
            .greet( "Hi", DEFAULT.withIdempotency("abcde"));
        // <end_service_idempotent>
    }

    public void attach() {

        // <start_service_attach>
        Client rs = Client.connect("http://localhost:8080");
        SendResponse handle = GreeterServiceClient.fromClient(rs)
                .send()
                .greet("Hi", DEFAULT.withIdempotency("abcde"));

        // ... do something else ...

        // Attach later to retrieve the result
        // withClass(1:3) highlight-line
        String greeting = rs
                .invocationHandle(handle.getInvocationId(), JsonSerdes.STRING)
                .attach();
        // <end_service_attach>
    }

    public void peekAtOutput(){
        // <start_service_peek>
        Client rs = Client.connect("http://localhost:8080");
        SendResponse handle = GreeterServiceClient.fromClient(rs)
                .send()
                .greet("Hi", DEFAULT.withIdempotency("abcde"));

        // ... do something else ...

        // Peek to see if the result is ready
        // withClass(1:3) highlight-line
        Output<String> output = rs
            .invocationHandle(handle.getInvocationId(), JsonSerdes.STRING)
            .getOutput();

        if (output.isReady()) {
            String result = output.getValue();
        }
        // <end_service_peek>
    }

    public void latchOntoWorkflow(){

        // <start_workflow_attach>
        Client rs = Client.connect("http://localhost:8080");
        SendResponse handle = MyWorkflowClient
                .fromClient(rs, "wf-id-1")
                .submit("input");

        // If you have access to the workflow handle:
        // withClass(1:3) highlight-line
        String response = rs
                .invocationHandle(handle.getInvocationId(), JsonSerdes.STRING)
                .attach();

        // If you do not have access to the workflow handle, use the workflow ID:
        String response2 = MyWorkflowClient.fromClient(rs, "wf-id-1")
                // withClass(1:2) highlight-line
                .workflowHandle()
                .attach();
        // <end_workflow_attach>

        // <start_workflow_peek>
        // If you have access to the workflow handle:
        // withClass(1:3) highlight-line
        Output<String> output = rs
                .invocationHandle(handle.getInvocationId(), JsonSerdes.STRING)
                .getOutput();

        if (output.isReady()) {
            String result = output.getValue();
        }

        // If you do not have access to the workflow handle, use the workflow ID:
        Output<String> output2 = MyWorkflowClient.fromClient(rs, "wf-id-1")
                // withClass(1:2) highlight-line
                .workflowHandle()
                .getOutput();
        // <end_workflow_peek>

    }
}
